### PR TITLE
Write the display flags in F entry when saving an annotation (issue 18072)

### LINKED ISSUE #18072: Fields not rendered after saving and opening Dutch tax form
Attach (recommended) or Link to PDF file here:

- Direct link: https://download.belastingdienst.nl/belastingdienst/docs/model_opgaaf_gegevens_loonheffingen_lh0082z11fol.pdf
- Web Archive link: https://web.archive.org/web/20240309103139/https://download.belastingdienst.nl/belastingdienst/docs/model_opgaaf_gegevens_loonheffingen_lh0082z11fol.pdf

Configuration:
- Web browser and its version: Firefox 125.0.3
- Operating system and its version: Arch Linux
- PDF.js version: 4.1.342 embedded in Firefox and 4.3.21 from https://mozilla.github.io/pdf.js/web/viewer.html
- Is a browser extension: no

Steps to reproduce the problem:
1. Open the PDF file (Dutch tax form).
2. Click the first radio button on the second page. Notice that extra fields appear behind it.
3. Enter a date: 12-05-2024
4. Save the PDF file to disk.
5. Open the saved PDF file. Notice that the first radio button on the second page is checked, but the extra fields behind it are no longer visible. This makes it _seem_ as if data loss occurred, even though that's not the case.
6. Click the second radio button below it and click the first radio button again. Notice that the extra fields behind the first radio button are now correctly visible, including the entered date.

What is the expected behavior? (add screenshot)
Expected

What went wrong? (add screenshot)
Actual

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension): https://mozilla.github.io/pdf.js/web/viewer.html

### PR DESCRIPTION
**Pull request by calixteman**
Fixes #18072.

### CODE REVIEW

File: src/core/annotation.js [hunk 1]
```diff
@@ -723,6 +723,38 @@ class Annotation {
     return !!(flags & flag);
   }
 
+  _buildFlags(noView, noPrint) {
```
Comment (Snuffleupagus) [inline src/core/annotation.js:726]: I've not review this method in detail, so I'm mostly trusting that you implemented all the flags correctly here :-)

File: src/core/annotation.js [hunk 2]
```diff
@@ -2089,7 +2126,8 @@ class WidgetAnnotation extends Annotation {
       Array.isArray(value) &&
       Array.isArray(this.data.fieldValue) &&
       value.length === this.data.fieldValue.length &&
-      value.every((x, i) => x === this.data.fieldValue[i])
+      value.every((x, i) => x === this.data.fieldValue[i]) &&
```
Comment (Snuffleupagus) [inline src/core/annotation.js:2129]: This is a pre-existing "issue", but there's a [helper function](https://github.com/mozilla/pdf.js/blob/783facb312c75941ba6bbfd5dffc52e85f46af5e/src/shared/util.js#L1017-L1027) we can use to simplify these two lines.
```javascript
isArrayEqual(value, this.data.fieldValue) &&
```

**Review by Snuffleupagus — APPROVED** [review]
r=me, with a small suggested simplification; thank you.

File: src/core/annotation.js (hunk 1 above)
Comment (calixteman) [reply src/core/annotation.js:726]: I set the F value to 0 and to 1023 in the pdf and then play with NoView/NoPrint in Acrobat, to see what bits are turned on/off, so at least it behaves the same way as in Acrobat.

File: src/core/annotation.js [hunk 3]
```diff
@@ -2089,7 +2127,8 @@ class WidgetAnnotation extends Annotation {
       Array.isArray(value) &&
       Array.isArray(this.data.fieldValue) &&
       value.length === this.data.fieldValue.length &&
```
Comment (Snuffleupagus) [inline src/core/annotation.js:2129]: Please note that this line is also handled by the `isArrayEqual` function.
```suggestion
```

**Comment by calixteman** [comment]
I cannot reproduce the windows issue locally but I wonder if it could be related to the fix for https://bugzilla.mozilla.org/show_bug.cgi?id=1919906.

**Comment by calixteman** [comment]
My guess about the potential culprit was correct and I filed a bug:
https://bugzilla.mozilla.org/show_bug.cgi?id=1922063

@Snuffleupagus, is it ok for you to make references even with this bug ?

**Comment by calixteman** [comment]
I changed the firefox binary on the windows bot in order to just the penultimate one.
/botio-windows test

**Comment by Snuffleupagus** [comment]
> is it ok for you to make references even with this bug ?

I don't think we have much choice, we'll "have" to do that for now, since otherwise testing other PRs become difficult if there's loads of unrelated failures.

**Pull request merged**